pkg/server/plugin/bootstrapper/k8sbootstrapper/k8sbootstrapper.go

```diff
@@ -0,0 +1,194 @@
+package k8sbootstrapper
+
+import (
+	"bytes"
+	"context"
+	"encoding/pem"
+	"net/http"
+	"sync"
+
+	"github.com/hashicorp/hcl"
+	spi "github.com/spiffe/spire/proto/common/plugin"
+	"github.com/spiffe/spire/proto/server/bootstrapper"
+	"github.com/zeebo/errs"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+var (
+	bootstrapErr = errs.Class("k8s-bootstrapper")
+)
+
+const (
+	defaultNamespace    = "spire"
+	defaultConfigMap    = "spire-bootstrap"
+	defaultConfigMapKey = "bootstrap.crt"
+)
+
+type ConfigUpdaterConfig struct {
+	Namespace          string `hcl:"namespace"`
+	ConfigMap          string `hcl:"config_map"`
+	ConfigMapKey       string `hcl:"config_map_key"`
+	KubeConfigFilePath string `hcl:"kube_config_file_path"`
+}
+
+type K8SBootstrapPlugin struct {
+	mu     sync.RWMutex
+	config *ConfigUpdaterConfig
+
+	hooks struct {
+		newClient func(configPath string) (kubeClient, error)
+	}
+}
+
+var _ bootstrapper.Plugin = (*K8SBootstrapPlugin)(nil)
+
+func New() *K8SBootstrapPlugin {
+	p := &K8SBootstrapPlugin{}
+	p.hooks.newClient = newClient
+	return p
+}
+
+func (p *K8SBootstrapPlugin) PublishBundle(stream bootstrapper.PublishBundle_PluginStream) error {
+	config, err := p.getConfig()
+	if err != nil {
+		return err
+	}
+
+	client, err := p.hooks.newClient(config.KubeConfigFilePath)
+	if err != nil {
+		return err
+	}
+
+	for {
+		// Get the config map so we can use the version to resolve conflicts racing
+		// on updates from other servers.
+		configMap, err := client.GetConfigMap(stream.Context(), config.Namespace, config.ConfigMap)
+		if err != nil {
+			return bootstrapErr.New("unable to get config map %s/%s: %v", config.Namespace, config.ConfigMap, err)
+		}
+
+		// Signal the caller that we are ready for the bundle
+		if err := stream.Send(&bootstrapper.PublishBundleResponse{}); err != nil {
+			return err
+		}
+
+		// Receive the bundle
+		req, err := stream.Recv()
+		if err != nil {
+			return err
+		}
+		if req.Bundle == nil {
+			return bootstrapErr.New("request missing bundle")
+		}
+		bundle := req.Bundle
+
+		// PEM encode all of the root CA certificates in the bundle
+		bundleData := new(bytes.Buffer)
+		for _, rootCA := range bundle.RootCas {
+			pem.Encode(bundleData, &pem.Block{
+				Type:  "CERTIFICATE",
+				Bytes: rootCA.DerBytes,
+			})
+		}
+
+		// Replace the config map data
+		configMap.Data = map[string]string{
+			config.ConfigMapKey: bundleData.String(),
```

Comment:
Would this overwrite any existing data in the config map aside from the `bootstrap.crt` key?

Reply:
I implemented patch first but then moved to update when some documentation lead me to believe the ResourceVersion-based conflict resolution didn't work with patch. I've found some more commentary suggesting it does work. I do think patch is the right operation, if it works with our conflict resolution strategy. I'll move back to patch.